src/sched/job_runner.cc

```diff
@@ -54,6 +58,98 @@ class JobRunnerWorker {
+    using job_queue_t = boost::lockfree::queue<job_t*>;
+
+    explicit JobRunnerStrand(std::weak_ptr<JobRunner> runner) : runner_weak_(runner) {}
+
+    bool AddJob(job_t&& job);
+
+    std::weak_ptr<JobRunner> runner_weak_;
+    HybridSpinMutex          hybrid_spin_mtx_;
+    bool                     has_ready_job_{false};
+    job_queue_t              pending_jobs_{kInitialQueueCapacity};
+
+    static constexpr std::size_t kInitialQueueCapacity = 8192;
+};
+
+// This function guarantees the safety and liveness of the strand.
+//
+// Safety:
+//   - There are no more than one jobs from the same strand in the bound runner;
+//   - Jobs are added to the runner in the order of being pushed to the pending queue.
+//
+// Liveness:
+//   - If there are jobs in the pending queue, there will be one of them being pushed
+//     to the runner, (even if there are no more incoming jobs).
+bool JobRunnerStrand::AddJob(JobRunnerStrand::job_t&& job) {
+    // Try to get the next job from the pending queue and run it.
+    //
+    // The function consists of 2 parts:
+    //   - Decision. It is protected by lock and it decides whether Push is needed.
+    //   - Push. It pushes the next job to the runner.
+    // Let's use + to denote `is_in_running_job == false` and `-` otherwise,
+    // use D to denote the D's that changed has_ready_job_ value and d to those not, then
+    //   - D/d+ only changes the has_ready_job_ value from false to true
+    //   - D/d- only changes the has_ready_job_ value from true to false
+    //   - P and D/d- has one-to-one matching and P always happen before the correspoding D/d
+    //   - D/d generates a P if and only if has_ready_job_ is true when exiting from D
+    //   - Then the D/d who generates a P will be a sequence like this:
+    //         [D+ d-* (D-)]*
+    //     D- does not generate a P but we add it as a placeholder.
+    //   - Then we will prove:
+    //       - in this sequence, the P generated by a D/d will always happen before next D/d
+    //       - If there is some D/d, its P does not happen before the next D/d:
+    //           - If the next D/d is a D/d-, since it must be generated by some P, and all previous P's have a
+    //             matched D/d-, there are not enough P's to generate this one, so it is impossible.
+    //           - If the next D/d is a D+, then the current one is D-, but D- does not generate P.
+    //   - So the sequence is like the following:
+    //         D+ [P d-] [P d-] [P D-] D+ [P d-] P ...
+    //     Jobs are denoted by brackets. Safety proved.
+    //
+    //  For liveness, D/d+ add jobs, D+ will push job itself so liveness is trivially hold, for d-, the job will
+    //  be visible to the next D/d. If we looks at the non-d- decision before it, it can only be D+/d-, and both
+    //  of them comes with a next D/d. Liveness proved.
+    auto try_to_run_next_job = [](JobRunnerStrand& strand, bool is_in_running_job) {
+        std::unique_ptr<job_t> next(nullptr);
```

Comment:
1. No need to init.
2. Usually for up/sp I use `= nullptr` to init in a more "pointer" style, it'll call the ctor anyway.